onnxruntime/core/graph/graph_utils.cc

```diff
@@ -809,5 +809,20 @@ NodeArg& CreateNodeArg(Graph& graph, const NodeArg& base_arg) {
   return graph.GetOrCreateNodeArg(graph.GenerateNodeArgName(base_arg.Name()), base_arg.TypeAsProto());
 }
 
+void ComputeConstantInitializerUseCount(const Graph& graph, std::unordered_map<std::string, size_t>& constant_initializers_use_count) {
+  for (const auto& node : graph.Nodes()) {
+    for (const auto* arg : node.InputDefs()) {
+      if (arg->Exists() && IsConstantInitializer(graph, arg->Name()))
+        constant_initializers_use_count[arg->Name()]++;
+    }
+
+    if (node.ContainsSubgraph()) {
+      for (const gsl::not_null<const Graph*>& subgraph : node.GetSubgraphs()) {
+        ComputeConstantInitializerUseCount(*subgraph, constant_initializers_use_count);
+      }
+    }
+  }
```

Comment:
The ONNX spec allows a value in a subgraph to shadow an outer scope value. If say the main graph had an initializer called X, and the subgraph also had an initializer called X, the current implementation will calculate a use count of 2, but they could contain completely different values so each should have a use count of 1.

Maybe this initial implementation could just check that an initializer in a subgraph is not shadowing an outer scope initializer (new augmented version of IsConstantInitializer that checks for shadowing by another initializer), and return a failure status if that is detected. I would guess that this happens rarely, but don't know for sure.

Reply:
Thanks Scott! This case only happens when X in both main graph and subgraph are constant initializers. The consequence is that X won't be released even it is packed. The model will run fine. I put the information down and made the function local.